### PR TITLE
Add RL on Verifiable Reward (RLVR) reference implementations

### PR DESCRIPTION
**Pull request by jacobthebanana**
This pull request includes:

- Minimalized TRL/VeRL to run one PPO optimization step, given token array, per-token advantage array, model checkpoint, and PPO hyperparameters.
- Agent SDK integration- define the environment using the familiar OpenAI Agent SDK and run RL on the LLM powering the agent. Not yet tested on multi-agent setups (agent as tool or handoff)
- Extensive typing for simplified function signatures and IDE support- static type checking, pyright lints, proper autocompletion even within the training loop.

Vec-Inf wrapper integration will come in a separate pull request.

### CODE REVIEW

**Comment by kohankhaki** [comment]
@jacobthebanana I’ll review this in multiple passes since it’s fairly large.
First I’ll focus on overall structure, then I’ll come back for details.

File: starters/llm_fine_tuning/README.md
Comment (kohankhaki) [inline starters/llm_fine_tuning/README.md]: The `starters/` directory no longer exists. Everything is now under `templates/`. 
RLVR is currently placed as a top-level directory (`templates/src/rlvr/`) alongside `llm/`, `mlp/`, `vlm/`. We need to have it under `rl/` directory. Since RLVR (Reinforcement Learning with Verifiable Reward) is a type of RL, it should be organized under `rl/` to match the documented structure.

File: templates/src/rlvr/grpo/train.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+            ExecutorConfig(
+                name="slurm",
+                executor=slurm_executor,
+                num_replicas=cfg.llm_judge_vllm.num_replicas,
+                concurrency=cfg.llm_judge_vllm.concurrency_per_replica,
+            ),
+        ]
+
+    def get_checkpoint_paths(self, step_index: int) -> _CheckpointPaths:
+        """Return checkpoint paths and recycle previous checkpoints if enabled.
```
Comment (kohankhaki) [inline templates/src/rlvr/grpo/train.py]: Can you please add a documentation on this?
It is a bit hard to understand the logic.

I think there is a problem with this:
- On step 0: `checkpoints = [base_model, base_model, new_checkpoint]` -> `_kl_ref=base_model, _current=base_model, _output=new_checkpoint` 
- On step 1: `checkpoints = [base_model, base_model, step_000, step_001]` -> `_kl_ref=base_model, _current=step_000, _output=step_001` 
- But the deletion logic at line 182-185 operates on `self.checkpoints` which doesn't include the base_model entries, so the indexing might be off.

Please let me know if I am wrong.

File: templates/src/rlvr/grpo/train.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            return asyncio.get_event_loop().run_until_complete(coro)
+
+    def calculate_advantage(
+        self,
+        data: Sequence[RLVRDataItem],
+        run_results: "Sequence[agents.RunResult]",
+        evals: Sequence[EvalResult],
+    ) -> AdvantageData:
+        """Calculate advantage given rollouts and eval results."""
+        tokenizer = AutoTokenizer.from_pretrained(self.cfg.base_model)
```
Comment (kohankhaki) [inline templates/src/rlvr/grpo/train.py]: Tokenizer is loaded fresh on every `calculate_advantage` call, which happens every training step. This is inefficient.

**Fix**: Load once in `__init__` and reuse.

**Review by kohankhaki — CHANGES_REQUESTED** [review]
I have left a few comments.
I might be wrong about a few of them, so please let me know if you think I am.

Below are the general comments:
- Several classes and helper functions lack docstrings. Please add them. :)

File: templates/src/rlvr/grpo/train.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+# Eval agent must support structured output and use output_type EvalResult
+eval_agent = agents.Agent(
+    "Evaluator",
+    instructions=(
+        "Evaluate if the `proposed` response matches the ground-truth `target`."
+        "Give a score of 0.0 if incorrect and 1.0 if correct."
+    ),
+    output_type=EvalResult,
+)
```
[the remark this answers is not in the document]
Comment (jacobthebanana) [reply templates/src/rlvr/grpo/train.py]: Are you sure about this? That will add a lot of new configurations (prompts, etc. for each agent,) and if the user want to add new tools to their agent, they will have to change code anyway

File: templates/src/rlvr/grpo/data_types.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        """Compute advantage for a given batch of rewards."""
+        if len(reward_details) == 0:
+            msg = "reward_details must be a non-empty list, but an empty list is given."
+            raise ValueError(msg)
+
+        # per-trace only, not token-level
+        rewards = np.asarray([_detail.reward for _detail in reward_details])
+        advantages_np: np.ndarray = np.asarray(
+            (rewards - rewards.mean()) / rewards.std()
+        )
```
[the remark this answers is not in the document]
Comment (jacobthebanana) [reply templates/src/rlvr/grpo/data_types.py]: Thanks for noticing this. And yes that did happen a few times during testing. The fix above makes sense.

File: templates/src/rlvr/grpo/grpo_frontend.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+def grpo_optimization_step(
+    advantage_data: AdvantageData,
+    current_policy_path: pathlib.Path,
+    kl_ref_path: pathlib.Path,
+    checkpoint_output_path: pathlib.Path,
+    hyperparameters: GRPOHyperparameters,
+    optimizer_path: pathlib.Path | None,
+) -> GRPOMetrics:
+    """Run one GRPO optimization step given advantages."""
+    device = torch.device("cuda:0")
```
[the remark this answers is not in the document]
Comment (jacobthebanana) [reply templates/src/rlvr/grpo/grpo_frontend.py]: Using the above to be consistent with other [files](https://github.com/VectorInstitute/vec-playbook/blob/ba0ea9bdcdc71ce5798bbfed2b2d67e26e519587/templates/src/mlp/single/train.py#L73-L74)

File: templates/src/rlvr/grpo/launch.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+"""Launch script for GRPO, Hydra + Submitit."""
+
+import logging
+import os
+
+import hydra
+from omegaconf import DictConfig, OmegaConf
+
+from templates.src.rlvr.grpo.config import GRPOConfig
+from templates.src.rlvr.grpo.train import GRPOTrainer
```
Comment (kohankhaki) [inline templates/src/rlvr/grpo/launch.py]: Please use relative imports through out `rlvr` template to match other templates.

File: templates/src/rl/rlvr/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+- Backprop GPU does not participate in rollouts.
+- Rollout GPUs might sit idle when all rollouts are done and only eval is pending.
+- Verify support for function calling via Chat Completion.
+- Integrate LangFuse datasets to track eval traces across steps.
+
+## Vaughan (Bon Echo) User? Extra setup steps required
+
+(Skip this section if you are running on Killarney)
+
+This reference implementation depends on `vllm>=0.11.0`. vLLM does not provide pre-built wheels for older Linux distros, such as the one running on Vaughan (Bon Echo). Follow the steps in [vaughan_setup.md](vaughan_setup.md) before continuing.
```
Comment (kohankhaki) [inline templates/src/rl/rlvr/README.md]: The README links to `[vaughan_setup.md](vaughan_setup.md)`, but the actual file in the directory is `bon_echo_setup.md`. The link will be broken. Please update line 27 to reference `bon_echo_setup.md` instead.

File: templates/src/rl/rlvr/grpo/config.yaml [hunk 8]
```diff
@@ -1,43 +1,34 @@
...
 
 vllm_cache_dir_base: "${oc.env:SCRATCH}/.cache/vllm_compiled_graphs/"
 
-base_vllm_config:
-  &base_vllm_config # # required for vaughan or if you installed vllm in a separate venv
-  submitit_python: "templates/configs/compute/vaughan/run_in_container.sh uv run python"
+base_vllm_config: &base_vllm_config
+  # (base config for reuse across rollout and judge)
+  # "submitit_python" is required for vaughan or if you installed vllm in a separate venv
+  submitit_python: "templates/configs/compute/bon_echo/run_in_container.sh uv run python"
```
Comment (kohankhaki) [inline templates/src/rl/rlvr/grpo/config.yaml]: `submitit_python` is hardcoded to Bon Echo container script. If Killarney is supported, please either:
  - make this cluster/compute-dependent, or
  - document that users must override it for Killarney.

File: templates/src/rl/rlvr/grpo/tests/test_submitit_vllm_local.py [hunk 9]
```diff
@@ -8,7 +8,8 @@
 from vllm.config import CompilationConfig
 from vllm.engine.arg_utils import EngineArgs
 
-from templates.src.rlvr.submitit_vllm import ExecutorConfig, SubmititVLLM
+# "best practice" to use absolute imports for pytest
+from templates.src.rl.rlvr.submitit_vllm import ExecutorConfig, SubmititVLLM
```
Comment (kohankhaki) [inline templates/src/rl/rlvr/grpo/tests/test_submitit_vllm_local.py]: This imports `from templates.src.rl.rlvr.submitit_vllm ...`. In this repo, `templates` is not a normal Python package (and `templates/src` is already the package root), so this path is likely fragile.
Other templates don’t ship tests under the template directory—consider moving integration tests elsewhere or dropping them from the template.
